### DISCUSSION TITLE
append to jsonb[defined as array] column on users table

### DISCUSSION THREAD

**Discussion by amrajat** [supabase/supabase#21091] [Questions] [answered]
I want to append the data into purchase_history column which is jsonb(not required for me  i can change the data type) which is defined as array.

i tried below code 
```
CREATE OR REPLACE FUNCTION add_purchase_history(user_id text, transaction_data jsonb)
RETURNS void AS $$
BEGIN
  UPDATE users
  SET purchase_history = purchase_history || transaction_data
  WHERE id = user_id::uuid;
END;
$$ LANGUAGE plpgsql;```

with rpc 
```
```
 let { data, error } = await supabase.rpc("add_purchase_history", {
    transaction_data: { planName: "Basic", qty: 10 },
    user_id: id,
  });
```

but i am getting this error
```
{
  code: 'PGRST202',
  details: 'Searched for the function public.add_purchase_history with parameters transaction_data, user_id or with a single unnamed json/jsonb parameter, but no matches were found in the schema cache.',
  hint: null,
  message: 'Could not find the function public.add_purchase_history(transaction_data, user_id) in the schema cache'
}
```

Thank you so much for your help! I really appreciate it!

**Comment by mwoss** [comment]
Have you tried to explicitly stringify the `transaction_data` parameters via `JSON.strinfigy()`. Not sure, but could help :3
Also, be sure that both values are not `undefined`, so the function is called with both parameters. Undefined parameters are skipped by default.

**Reply by amrajat** [reply]
Also tired with stringify, parameters aren't undefined.  Sometimes it throws above error. Sometimes it fails silently.

**Reply by mwoss** [reply]
When you try to call it from the SQL editor it work?

**Reply by amrajat** [reply]
select add_purchase_history('dfac2d1a-50c9-4285', '{"planName": "Basic", "qty": 10}'); is working fine in SQL editor. but it's not working with rpc??? I must be doing something stupid.

**Comment by SaddamMohsen** [comment]
the problem come from this parameter name (planName)
you must change it to lower case planname or plan_name 
when you send data to api it converts all name to lower case

**Reply by amrajat** [reply]
I tried, but it didn't work.

**Reply by GaryAustin1** [reply]
The jsonb names have no part in the error.

**Reply by amrajat** [reply]
It's working fine with sql editor, not on rpc

**Reply by SaddamMohsen** [reply]
if it is working in sql editor 
then do what i said to change all the parameter name to lower case in function declaration and in the invocation

**Reply by amrajat** [reply]
already tried but it's not working. it not updating my tables.

**Reply by SaddamMohsen** [reply]
maybe there is something related to policies

**Comment by GaryAustin1** [comment]
Try flushing the PostgREST cache.  Maybe you changed the parameters at some point... (but that does not seem right based on error message).
In the SQL editor run `NOTIFY pgrst, 'reload schema'`
I just ran:
```sql
CREATE OR REPLACE FUNCTION addp(user_id text, transaction_data jsonb)
RETURNS void AS $$
BEGIN
raise log 'user_id = %',user_id;
END;
$$ LANGUAGE plpgsql;
```
And:
```js
    const data = await supabase
        .rpc('addp',{transaction_data:{Cat:'cat'},user_id:'cat'})
```
  Which works.

**Reply by GaryAustin1** [reply]
Is the function in the public schema?
Are you going to the correct instance?

Edit:  Is that your EXACT function code including all lower case letters for the name and the parameter names (not counting the jsonb key name)?

**Reply by amrajat** [reply]
the function is in the public schema. I tried to reload the schema and restart my project but didn't work.
Edit: Now it's giving me null it's not throwing the PGRST202 error anymore.???

**Reply by GaryAustin1** [reply]
null is good as it is a void function.

![image](https://github.com/supabase/supabase/assets/54564956/94719919-3a29-4961-a686-65b48188f6df)

**Reply by GaryAustin1** [reply]
Did you edit the parameter names, types, or add or take away a parameter?
If so and you did not flush the cache it was likely confused.

**Reply by GaryAustin1** [reply]
If you are saying it is not updating your table then that is likely RLS, or a bug in that part of your function.

**Reply by amrajat** [reply]
the logs are also saying it's success 204 but it's not showing in my tables.

**Reply by GaryAustin1** [reply]
See answer above.  You have to meet RLS for select and update to do update.

**Reply by amrajat** [reply]
RLS is active for users and I am testing as a user.  so RLS is not the issue here.

INSERT

Users can insert their own profile.

Applied to:

public


UPDATE

Users can update own profile.

Applied to:

public


SELECT

Users can view their own profile.

Applied to:

public

**Reply by GaryAustin1** [reply]
Turn off RLS to test.
You can also simulate being a user in the SQL editor bottom right.
You are not showing policies just labels.
Many cases of users thinking they have a session when they make the call and they don't.

RLS is silent for update so will not error.

**Reply by amrajat** [reply]
log are saying /rest/v1/rpc/add_purchase_history is 204 which is a success. but it's not updating my tables.

**Reply by GaryAustin1** [reply]
Turn off RLS as a test.  There is no error for not meeting RLS.  It will return fine and not do the update.

**Reply by amrajat** [reply]
i disabled the RLS and it's logging the null as before without any new error.

**Reply by GaryAustin1** [reply]
It will be null.  You don't return anything.  You have a void for the return.
Did it update now?

**Reply by amrajat** [reply]
yes it did. but why?

**Reply by GaryAustin1** [reply]
Because you don't meet RLS.

**Reply by amrajat** [reply]
USING expression=true, 
WITH CHECK expression=(auth.uid() = id) isn't it correct?

**Reply by GaryAustin1** [reply, marked as answer]
You can show your policies (not just the labels).  But it could also be you don't have a user session when you make the rpc call in the supabase client making the call.

You can test if you have a session by turning back on RLS and making your select and update policies both true and the role option anon instead of public.  If that works then you don't have a user session.  Note you should also make sure your update policy has the USING part set.

**Reply by GaryAustin1** [reply]
Is that your update policy or your select and update policy?
You need to meet select policy.  And you need to meet update policy.  Update policy must have the USING part, WITH CHECK is optional.

**Reply by amrajat** [reply]
Once I had a similar issue with RLS then I updated the RLS. I apologize for not showing the policies. Thank you so much for your time and efforts. I appreciate it.

**Reply by amrajat** [reply]
I have this rpc function in my stripe webhook, when a payment happens on stripe, the stripe calls this webhook, which uses the rpc and updates the data. So technically there is no session saved on the webhook, how can I set the policies accordingly?

**Reply by GaryAustin1** [reply]
Then use service role for your client in the webhook so it bypasses RLS.  I assume your webhook is running on a secure server environment if it is dealing with Stripe stuff.
